packages/composite-checkout/src/components/coupon.js

```diff
@@ -0,0 +1,203 @@
+			<Field
+				id={ id }
+				placeholder={ localize( 'Enter your coupon code' ) }
+				isError={ hasCouponError }
+				errorMessage={
+					hasCouponError
+						? localize( "We couldn't find your coupon. Please check your code and try again." )
+						: null
+				}
+				onChange={ input => {
+					handleFieldInput( input, setCouponFieldValue, setIsApplyButtonActive, setHasCouponError );
+				} }
+			/>
+
+			{ isApplyButtonActive && (
+				<ApplyButton
+					buttonState="secondary"
+					onClick={ value => {
+						handleApplyButtonClick(
+							value,
+							setHasCouponError,
+							couponAdded,
+							setIsCouponApplied,
+							setIsNoticeVisible
+						);
+					} }
+				>
+					{ localize( 'Apply' ) }
+				</ApplyButton>
+			) }
+
+			{ isNoticeVisible && <NoticeComponent /> }
+		</CouponWrapper>
+	);
+}
+
+Coupon.propTypes = {
+	id: PropTypes.string.isRequired,
+	couponAdded: PropTypes.func,
+};
+
+const animateIn = keyframes`
+  from {
+    opacity: 0;
+    transform: translateX(8px);
+  }
+
+  to {
+    opacity: 1;
+    transform: translateY(0);
+  }
+`;
+
+const CouponWrapper = styled.div`
+	margin: ${props => props.marginTop} 0 0 0;
+	padding-top: 0;
+	position: relative;
+	width: 100%;
+`;
+
+const ApplyButton = styled( Button )`
+	position: absolute;
+	top: 4px;
+	right: 4px;
+	padding: 7px;
+	animation: ${animateIn} 0.2s ease-out;
+`;
+
+function NoticeComponent() {
+	const localize = useLocalize();
+
+	return (
+		<Notice type="success">
+			{ localize( 'Your coupon for $20 has been applied to your cart' ) }
+		</Notice>
+	);
+}
+
+function handleApplyButtonClick(
+	value,
+	setHasCouponError,
+	couponAdded,
+	setIsCouponApplied,
+	setIsNoticeVisible
+) {
+	//TODO: Validate coupon field and replace couponAdded in the following if statement
+	if ( couponAdded ) {
+		setIsNoticeVisible( true );
+		setIsCouponApplied( true );
+		setTimeout( () => {
+			setIsNoticeVisible( false );
+		}, 2000 );
+
+		if ( couponAdded ) {
+			couponAdded();
+		}
+
+		return;
+	}
+
+	setHasCouponError( true );
+	return;
+}
+
+function handleFieldInput( input, setCouponFieldValue, setIsApplyButtonActive, setHasCouponError ) {
+	if ( input.length > 0 ) {
+		setCouponFieldValue( input );
+		setIsApplyButtonActive( true );
+		setHasCouponError( false );
+		return;
+	}
+
+	setIsApplyButtonActive( false );
+	return;
+}
+
+function useCouponField(
+	id,
+	couponFieldValue,
+	setHasCouponError,
+	couponAdded,
+	setIsCouponApplied,
+	setIsNoticeVisible
+) {
+	useEffect( () => {
+		const keyPressHandler = makeHandleKeyPress(
+			couponFieldValue,
+			setHasCouponError,
+			couponAdded,
+			setIsCouponApplied,
+			setIsNoticeVisible
+		);
+		if ( document.getElementById( id ) === document.activeElement ) {
+			document.addEventListener( 'keydown', keyPressHandler, false );
+		}
+
+		return () => document.removeEventListener( 'keydown', keyPressHandler, false );
+	}, [ couponFieldValue ] );
```

Comment:
I'm surprised the linter doesn't complain that `couponAdded` and `id` are not included in the dependencies array. I think that if you don't include them here then the values will be out-of-date when the keypress happens.

Also... can you explain more about the `document.activeElement` check? Remember that this function will only run when the component re-renders and one of the dependencies has changed. Since `activeElement` isn't a dependency, if it changes separately from the dependencies, this listener will not get added (or removed).

🤔 If the purpose of all of this is to allow the enter key to activate the coupon field when it's focused, there's probably a better way. If we wrap the coupon `input` with a `form` tag and use the form's `onSubmit` attribute, we get the "submit on enter" behavior for free.

Reply:
> can you explain more about the document.activeElement check?

yeah, you're right. This was to check if the field was being focused. You suggestion of using the form sounds like a way better idea! I didn't think of that.